File: examples/vulnerable/csharp/PathTraversal.cs
/**
 * DELIBERATELY VULNERABLE CODE — FOR EDUCATIONAL / GHAS TESTING PURPOSES ONLY
 *
 * Vulnerability : Path Traversal (Directory Traversal)
 * CWE           : CWE-22  (Improper Limitation of a Pathname to a Restricted Directory)
 * CodeQL query  : cs/path-injection
 *
 * Description:
 *   A user-supplied filename is combined with a base directory using
 *   Path.Combine and then read without any validation.  An attacker can
 *   supply values such as "../../etc/passwd" or "..\..\windows\win.ini"
 *   to escape the intended directory and read arbitrary files on the
 *   server's filesystem.
 *
 * Remediation:
 *   - Validate that the resolved canonical path starts with the intended
 *     base directory (Path.GetFullPath + StartsWith check).
 *   - Reject filenames that contain ".." or path separator characters.
 *   - Use a whitelist of allowed filenames where possible.
 */

using System.IO;
using Microsoft.AspNetCore.Mvc;

namespace GhasPack.Vulnerable.CSharp
{
    [ApiController]
    [Route("api/[controller]")]
    public class PathTraversalController : ControllerBase
    {
        private const string UploadsDirectory = "/uploads";

        // GET /api/pathtraversal?file=report.pdf
        [HttpGet]
        public IActionResult DownloadFile([FromQuery] string file)
        {
            // BAD: user-controlled filename used without validation
            string filename = file;

            // BAD: Path.Combine does NOT prevent traversal sequences
            string filePath = Path.Combine(UploadsDirectory, filename);

            if (!System.IO.File.Exists(filePath))
            {
                return NotFound("File not found.");
            }

            // BAD: arbitrary file read — attacker can traverse outside /uploads
            string content = System.IO.File.ReadAllText(filePath);

            return Ok(content);
        }

        // GET /api/pathtraversal/download?file=report.pdf
        [HttpGet("download")]
        public IActionResult DownloadFileBytes([FromQuery] string file)
        {
            // BAD: same pattern, returning raw bytes
            string filePath = Path.Combine(UploadsDirectory, file);

            // BAD: no canonical-path check before reading
            byte[] data = System.IO.File.ReadAllBytes(filePath);

            return File(data, "application/octet-stream", file);
        }
    }
}

File: examples/vulnerable/python/path_traversal.py
"""
Path Traversal — CWE-22
https://cwe.mitre.org/data/definitions/22.html

⚠️  DELIBERATELY VULNERABLE — Do not use in production.

This file demonstrates path traversal in a Flask application where a
user-controlled filename is used to construct a file path without
validation.  An attacker can supply:

    GET /download?file=../../../etc/passwd

to read arbitrary files on the server.

Detected by: CodeQL (python/path-injection)
Fixed version: ../../secure/python/path_traversal.py
"""

import os
from flask import Flask, request, send_file, jsonify

app = Flask(__name__)

UPLOAD_DIR = "/uploads"


# BAD: User-controlled file path without validation
@app.route("/download")
def download():
    filename = request.args.get("file", "")
    file_path = os.path.join(UPLOAD_DIR, filename)
    return send_file(file_path)


# BAD: User input used to construct path with string formatting
@app.route("/view")
def view_file():
    filename = request.args.get("file", "")
    file_path = f"/uploads/{filename}"

    with open(file_path, "r") as f:
        content = f.read()

    return jsonify({"filename": filename, "content": content})


# BAD: User input in directory listing
@app.route("/list")
def list_files():
    directory = request.args.get("dir", "")
    full_path = os.path.join(UPLOAD_DIR, directory)

    files = os.listdir(full_path)
    return jsonify({"directory": directory, "files": files})


if __name__ == "__main__":
    app.run(debug=True, port=5000)
